retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofitBuilder.java

```diff
@@ -0,0 +1,221 @@
+ * under the License.
+ */
+package com.linecorp.armeria.client.http.retrofit2;
+
+import java.net.URI;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.client.http.HttpClient;
+import com.linecorp.armeria.common.Scheme;
+import com.linecorp.armeria.common.SessionProtocol;
+
+import okhttp3.HttpUrl;
+import retrofit2.CallAdapter;
+import retrofit2.Converter;
+import retrofit2.Retrofit;
+import retrofit2.Retrofit.Builder;
+
+/**
+ * A helper class for creating a new {@link Retrofit} instance with {@link ArmeriaCallFactory}.
+ * For example,
+ * <pre>{@code
+ *
+ * Retrofit retrofit = new ArmeriaRetrofitBuilder("none+http://localhost:8080/")
+ *     .build();
+ *
+ * MyApi api = retrofit.create(MyApi.class);
+ * Response&lt;User&gt; user = api.getUser().execute();
+ * }
+ * </pre>
+ *
+ * <p>ArmeriaRetrofitBuilder even supports EndpointGroup, so you can create Retrofit like below,
+ * <pre>{@code
+ *
+ * EndpointGroupRegistry.register("foo",
+ *                                new StaticEndpointGroup(Endpoint.of("127.0.0.1", 8080)),
+ *                                ROUND_ROBIN);
+ *
+ * Retrofit retrofit = new ArmeriaRetrofitBuilder("none+http://group:foo/")
+ *     .build();
+ * }
+ * </pre>
+ * ArmeriaRetrofitBuilder will convert http://group:foo to http://__group__foo internally to avoid okHttp3's
+ * {@link HttpUrl} throwing exception when parsing http://group:foo . To avoid your client have possibility to
+ * access domain which starts with __group__, you can use {@link ArmeriaRetrofitBuilder#groupPrefix(String)} to
+ * customize the internal behavior.
+ *
+ * <p>If you want to decorate HttpClient, using {@link ArmeriaRetrofitBuilder#newClientFunction(Function)} to
+ * customize.
+ * <pre>{@code
+ *
+ * Retrofit retrofit = new ArmeriaRetrofitBuilder("none+http://localhost:8080/")
+ *     .newClientFunction(uri -> new ClientBuilder(uri)
+ *             .decorator(HttpRequest.class, HttpResponse.class, LoggingClient::new)
+ *             .build(HttpClient.class))
+ *     .build();
+ * }
+ * </pre>
+ */
+public final class ArmeriaRetrofitBuilder {
+
+    private static final Pattern GROUP_PREFIX_PATTERN = Pattern.compile("^[_0-9a-z]+$");
+    private static final String DEFAULT_GROUP_PREFIX = "__group__";
+    private static final Function<String, HttpClient> DEFAULT_NEW_CLIENT_FUNCTION =
+            url -> Clients.newClient(ClientFactory.DEFAULT, url, HttpClient.class);
+    public static final String SLASH = "/";
+
+    private final Retrofit.Builder retrofitBuilder;
+    private String baseUrl;
+    private HttpClient baseHttpClient;
+    private String basePath;
+    private String groupPrefix = DEFAULT_GROUP_PREFIX;
+    private Function<String, HttpClient> newClientFunction = DEFAULT_NEW_CLIENT_FUNCTION;
```

Comment:
I want to let user can know which url they will be create, hmm. maybe `BiFunction<URL, ? super ClientOptions, ClientOptions>` ?